src/libbson/src/bson/bson-utf8.c

```diff
@@ -266,54 +288,106 @@ bson_utf8_escape_for_json (const char *utf8, /* IN */
    bson_unichar_t c;
    bson_string_t *str;
    bool length_provided = true;
-   const char *end;
+   ssize_t pos;
 
    BSON_ASSERT (utf8);
 
-   str = bson_string_new (NULL);
-
    if (utf8_len < 0) {
       length_provided = false;
       utf8_len = strlen (utf8);
    }
 
-   end = utf8 + utf8_len;
+   if (utf8_len == 0) {
+      return bson_strdup ("");
+   }
+
+   str = bson_string_alloc (utf8_len);
+
+   pos = 0;
+
+   do {
+      c = (unsigned char) utf8[pos];
```

Comment:
Is the input to `bson_utf8_escape_for_json` expected to be valid UTF-8?

This test results in an error reported by ASAN:

```c
str = bson_utf8_escape_for_json ("\xc6", 1);
ASSERT_MEMCMP (str, "\\u00c6", 2);
bson_free (str);
```

Error:
```
==25574==ERROR: AddressSanitizer: global-buffer-overflow on address 0x000100f51942 at pc 0x000100ebc978 bp 0x00016f97a2b0 sp 0x00016f97a2a8
READ of size 1 at 0x000100f51942 thread T0
    #0 0x100ebc974 in bson_utf8_escape_for_json bson-utf8.c:309
    #1 0x10080b540 in test_bson_utf8_escape_for_json test-utf8.c:87
    #2 0x100d32d54 in TestSuite_AddHelper TestSuite.c:261
    #3 0x100d383d0 in TestSuite_RunTest TestSuite.c:648
    #4 0x100d35b9c in TestSuite_RunAll TestSuite.c:1094
    #5 0x100d34a48 in TestSuite_Run TestSuite.c:1139
    #6 0x100486da0 in main test-libmongoc-main.c:157
    #7 0x101d85088 in start+0x204 (dyld:arm64+0x5088) (BuildId: 6c9a46d971373785afc97feeceb8cadb32000000200000000100000000070c00)
    #8 0x12597ffffffffffc  (<unknown module>)
```

If this `utf8` is required to be valid UTF-8, I suggest updating the documentation to note.

Reply:
> Is the input to `bson_utf8_escape_for_json` expected to be valid UTF-8?

I'd say "yes". In the context of JSON encoding, for which this function is written, the `bson_iter_visit_all` function always calls `bson_utf8_validate` before invoking the respective visitor function, which is where `bson_utf8_escape_for_json` is called. With that in mind, I don't think we need to handle invalid UTF-8, but I defer to your judgment on this.

Reply:
Prior to this PR, the behavior of encoding `\xc6` appears to produce incorrect results:

```c
   // Test invalid UTF-8.
   // Hex C6 represents the bits 11000110.
   // The leading bits 110 signals a sequence of two bytes.
   // But only one byte is in the input.
   str = bson_utf8_escape_for_json ("\xc6", 1);
   BSON_ASSERT (!str); // Expected NULL returned?
   // Prior to PR, results in encoded value C680.
   // With PR, results in global-buffer-overflow.
```

I think it is reasonable to assume `bson_utf8_escape_for_json` has undefined behavior for invalid UTF-8 input. I suggest updating the documentation in `bson_utf8_escape_for_json.rst` to warn:

```rst
.. warning::
  
  If ``utf8`` represents invalid UTF-8 data the behavior of :symbol:`bson_utf8_escape_for_json` is undefined.
  UTF-8 may be validated with :symbol:`bson_utf8_validate`.
```